core/src/test/java/org/elasticsearch/snapshots/BlobStoreFormatTests.java

```diff
@@ -0,0 +1,248 @@
+        randomCorruption(blobContainer, "test-path");
+        try {
+            checksumFormat.read(blobContainer, "test-path");
+            fail("Should have failed due to corruption");
+        } catch (ElasticsearchCorruptionException ex) {
+            assertThat(ex.getMessage(), startsWith("[test-path] "));
+        } catch (EOFException ex) {
+            // This can happen if corrupt the byte length
+        }
+    }
+
+    public void testAtomicWrite() throws Exception {
+        final BlobStore blobStore = createTestBlobStore();
+        final BlobContainer blobContainer = blobStore.blobContainer(BlobPath.cleanPath());
+        String testString = randomAsciiOfLength(randomInt(10000));
+        final CountDownLatch block = new CountDownLatch(1);
+        final CountDownLatch unblock = new CountDownLatch(1);
+        final BlobObj blobObj = new BlobObj(testString) {
+            @Override
+            public void toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+                super.toXContent(builder, params);
+                // Block before finishing writing
+                try {
+                    block.countDown();
+                    unblock.await(5, TimeUnit.SECONDS);
+                } catch (InterruptedException ex) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        };
+        final TestSimpleBlobStoreFormat checksumFormat = new TestSimpleBlobStoreFormat(true, randomBoolean() ? XContentType.SMILE : XContentType.JSON);
+        ExecutorService threadPool = Executors.newFixedThreadPool(1);
+        try {
+            Future<Void> future = threadPool.submit(new Callable<Void>() {
+                @Override
+                public Void call() throws Exception {
+                    checksumFormat.writeAtomic(blobObj, blobContainer, "test-blob", "temp-test-blob", randomBoolean());
+                    return null;
+                }
+            });
+            block.await(5, TimeUnit.SECONDS);
+            assertFalse(blobContainer.blobExists("test-blob"));
+            unblock.countDown();
+            future.get();
+            assertTrue(blobContainer.blobExists("test-blob"));
+        } finally {
+            threadPool.shutdown();
+        }
+    }
+
+    protected BlobStore createTestBlobStore() throws IOException {
+        Settings settings = Settings.builder().build();
+        return new FsBlobStore(settings, randomRepoPath());
+    }
+
+    protected void randomCorruption(BlobContainer blobContainer, String blobName) throws IOException {
+        byte[] buffer = new byte[(int)blobContainer.listBlobsByPrefix(blobName).get(blobName).length()];
+        try (InputStream inputStream = blobContainer.openInput(blobName)) {
+            Streams.readFully(inputStream, buffer);
+        }
+        int location = randomIntBetween(0, buffer.length - 1);
+        buffer[location] = (byte)(buffer[location] ^ 42);
```

Comment:
can we compute CRC on `buffer` before and after introducing this corrupted byte. We can assume() the checksum has actually changed, so the test does not fail rarely?